Initial Multichain API docs (#1621)
MetaMask/metamask-docs: wallet/concepts/multichain-api.md

---
description: Learn about the Multichain API.
sidebar_custom_props:

>   flask_only: true

Reviewer: it's flask and beta soon i believe. @vandan

Reviewer 2: Yes, let's remove the Flask label. We can just make a note that the APIs are considered experimental and may be subject to changes.

Reviewer: Since we're on our way to hitting prod soon. Let's do no flask label, no beta label, just an experimental label. I think that's what Vandan is saying

---

# About the Multichain API

:::flaskOnly
:::

The Multichain API is a scalable, generalized web3 wallet API that supports simultaneous

> interactions across multiple blockchain networks.

Reviewer: the terms network and ecosystems are overloaded, but is it clearer to mention `networks and ecosystems`? I take networks to mean chains inside an ecosystem

Author: There was a thread about "network" terminology in Slack [here](https://consensys.slack.com/archives/C03Q3TCUVRT/p1743528144279929). I think the consensus was to continue to use "networks" for end users, but developers have more bandwidth to understand EVM vs. non-EVM. I think using "networks and ecosystems" here makes sense, added :+1:

authorization scopes.
Dapps can update the connection using the same method `wallet_createSession`.

Dapps can use [`wallet_invokeMethod`](../reference/multichain-api.md#wallet_invokemethod) to
call a subset of the [Wallet JSON-RPC API methods](../reference/json-rpc-methods/index.md) on
a specified chain.
Dapps can use [`wallet_getSession`](../reference/multichain-api.md#wallet_getsession) to get
the scopes and properties of the active connection, and use
[`wallet_revokeSession`](../reference/multichain-api.md#wallet_revokesession) to revoke the connection.
The API also supports the following events:

- [`wallet_notify`](../reference/multichain-api.md#wallet_notify) notifies dapps of onchain events or state changes they previously subscribed to.
- [`wallet_sessionChanged`](../reference/multichain-api.md#wallet_sessionchanged) notifies dapps of changes to the multichain connection.

See the [Multichain API reference](../reference/multichain-api.md) for full details.

### Lifecycle diagram

The following sequence diagram illustrates the multichain connection lifecycle.

```mermaid
%%{
  init: {
    'sequence': {
      'actorMargin': 100,
      'width': 275
    }
  }
}%%

> sequenceDiagram

Reviewer: do we need to show the case where sessionChange is successful?

Author: This diagram was taken from [CAIP 316](https://github.com/ChainAgnostic/CAIPs/blob/main/CAIPs/caip-316.md#visualizing-the-lifecycle-of-a-session-without-a-sessionid). Would a successful sessionChange case look the same as the failed case, minus the "connection interrupted" / "connection re-established" arrows?

Reviewer: oh. well... let's just leave it then haha

  different wallet APIs and SDKs, making it easier for dapps to integrate with wallets, discover
  capabilities, and negotiate interfaces.

[**Get started using the Multichain API.**](../how-to/manage-networks/use-multichain.md)

## Technical overview

The Multichain API follows the [CAIP-25](https://github.com/ChainAgnostic/CAIPs/blob/main/CAIPs/caip-25.md)
standard for dapps to interface with multichain wallets.
The API includes a method [`wallet_createSession`](../reference/multichain-api.md#wallet_createsession)
that dapps can call to create a multichain session with a wallet, with specified properties and
authorization scopes.
Dapps can update the session using the same method `wallet_createSession`.

Dapps can use [`wallet_invokeMethod`](../reference/multichain-api.md#wallet_invokemethod) to
call a subset of the [Wallet JSON-RPC API methods](../reference/json-rpc-methods/index.md) on
a specified chain.
Dapps can use [`wallet_getSession`](../reference/multichain-api.md#wallet_getsession) to get
the scopes and properties of the active session, and use
[`wallet_revokeSession`](../reference/multichain-api.md#wallet_revokesession) to revoke the session.

> The API also supports the [`wallet_notify`](../reference/multichain-api.md#wallet_notify) and

Reviewer 2: `wallet_notify` is a little different in that it provides notifications related to on-chain updates or state changes that the dapp has subscribed to. Whereas, `wallet_sessionChanged` only provides notifications about session or "permission" changes.

  init: {
    'sequence': {
      'actorMargin': 100,
      'width': 275
    }
  }
}%%

sequenceDiagram
  participant Dapp
  participant Wallet
  participant WalletDataStore as Wallet data store
  
  opt Create connection
  Dapp->>Wallet: wallet_createSession
  Wallet->>WalletDataStore: Persist connection data
  Wallet-->>Dapp: {"sessionScopes": {...}}
  end
  
  opt Update connection
  Dapp->>Wallet: wallet_createSession (update auth)
  Wallet->>WalletDataStore: Update connection data
  Wallet-->>Dapp: {"sessionScopes": {updatedScopes...}}
  end
  
  opt Connection interrupted with wallet-side modification
  Dapp-->>Wallet: Connection interrupted
  Wallet->>WalletDataStore: User initiated connection change
  Wallet-->>Dapp: wallet_sessionChanged (attempt fails)
  Dapp-->>Wallet: Connection re-established
  end
  
  opt Get connection
  Dapp->>Wallet: wallet_getSession
  Wallet-->>Dapp: {"sessionScopes": {...}}
  end

  opt Revoke connection
  Dapp->>Wallet: wallet_revokeSession
  Wallet->>WalletDataStore: Update connection data
  Wallet-->>Dapp: {"result": "true"}
  end
```

## Backwards compatibility

When using the Multichain API, your dapp can still interact with the existing

> [Wallet Provider API](wallet-api.md#ethereum-provider-api).

Reviewer 2: ```suggestion
[Ethereum Provider API](wallet-api.md#ethereum-provider-api).
```

> However, the Provider API is not optimized for multichain, and we recommend

Reviewer 2: ```suggestion
However, the Ethereum Provider API is not optimized for multichain usage, and we recommend
```

[MIP-6](https://github.com/MetaMask/metamask-improvement-proposals/blob/main/MIPs/mip-6.md) for
detailed information about MetaMask's Multichain API implementation.

The API includes a method [`wallet_createSession`](../reference/multichain-api.md#wallet_createsession)
that dapps can call to create a multichain connection with a wallet, with specified properties and
authorization scopes.
Dapps can update the connection using the same method `wallet_createSession`.

Dapps can use [`wallet_invokeMethod`](../reference/multichain-api.md#wallet_invokemethod) to
call a subset of the [Wallet JSON-RPC API methods](../reference/json-rpc-methods/index.md) on
a specified chain.
Dapps can use [`wallet_getSession`](../reference/multichain-api.md#wallet_getsession) to get
the scopes and properties of the active connection, and use
[`wallet_revokeSession`](../reference/multichain-api.md#wallet_revokesession) to revoke the connection.
The API also supports the following events:

- [`wallet_notify`](../reference/multichain-api.md#wallet_notify) notifies dapps of onchain events or state changes they previously subscribed to.

> - [`wallet_sessionChanged`](../reference/multichain-api.md#wallet_sessionchanged) notifies dapps of changes to the multichain connection.

Reviewer 3: Perhaps we can improve readability here?
Maybe something like, "The API incudes methods for managing multichain connections between dapps and wallets:", then list `wallet_createSession`, `wallet_invokeMethod` in a list format?

```mermaid
%%{
  init: {
    'sequence': {
      'actorMargin': 100,
      'width': 275
    }
  }
}%%

sequenceDiagram
  participant Dapp
  participant Wallet
  participant WalletDataStore as Wallet data store
  
  opt Create connection
  Dapp->>Wallet: wallet_createSession
  Wallet->>WalletDataStore: Persist connection data
  Wallet-->>Dapp: {"sessionScopes": {...}}
  end
  
  opt Update connection
  Dapp->>Wallet: wallet_createSession (update auth)
  Wallet->>WalletDataStore: Update connection data
  Wallet-->>Dapp: {"sessionScopes": {updatedScopes...}}
  end
  
  opt Connection interrupted with wallet-side modification
  Dapp-->>Wallet: Connection interrupted
  Wallet->>WalletDataStore: User initiated connection change
  Wallet-->>Dapp: wallet_sessionChanged (attempt fails)
  Dapp-->>Wallet: Connection re-established
  end
  
  opt Get connection
  Dapp->>Wallet: wallet_getSession
  Wallet-->>Dapp: {"sessionScopes": {...}}
  end

  opt Revoke connection
  Dapp->>Wallet: wallet_revokeSession
  Wallet->>WalletDataStore: Update connection data
  Wallet-->>Dapp: {"result": "true"}
  end
```

## Backwards compatibility

> When using the Multichain API, your dapp can still interact with the existing

Reviewer 3: Should we update the Wallet API topic to make users aware of this?

[Ethereum provider API](wallet-api.md#ethereum-provider-api).
However, the provider API is not optimized for multichain usage, and we recommend
[starting directly with the Multichain API](../how-to/manage-networks/use-multichain.md).
The Multichain API is backwards compatible mainly to support dapps that use third-party libraries

> with dependencies on the legacy provider.

Reviewer: not necessarily in this section, but is it worth noting that the EIP-1193 and Multichain sessions are the same session and that one will overwrite the other? I'm sure this will be a point of confusion for developers @vandan @adonesky1

Author: Added a note to this section, lmk if you'd like wording changes.

Reviewer 2: Agreed. Thank you.

Reviewer 4: We'll probably want to re-visit this since we may want to add mitigations for this overwriting case

---
description: Learn about the Multichain API.
---

# About the Multichain API

:::tip Experimental
The Multichain API is an experimental feature and currently only available on the extension (not on mobile).
:::

The Multichain API is a scalable, generalized web3 wallet API that supports simultaneous
interactions across multiple blockchain networks and ecosystems.
When integrated with [MetaMask Snaps](/snaps), it enables developers to interact with both popular
and emerging networks.
Key benefits include:

> - **Elimination of chain switching** - The Multichain API allows dapps to interact with multiple EVM networks without having to request chain switches.

Reviewer 4: inclusion of `EVM` here seems to imply the concurrency is only available for EVM networks
```suggestion
- **Elimination of chain switching** - The Multichain API allows dapps to interact with multiple networks without having to request chain switches.
```

Reviewer 2: Right, but chain switching is primarily an EVM specific concern.

>   This feature significantly reduces development overhead involved with error handling.

Reviewer 4: ```suggestion
  This feature significantly reduces development overhead involved with ensuring the correct network is targeted for a given dapp proposed transaction.
```

> - **Extensibility** - The Multichain API can be integrated with
>   [interoperability Snaps](https://snaps.metamask.io/explore/), providing a standards-based interface
>   to non-EVM networks that can be broadly adopted across ecosystems.

Reviewer 4: I would recommend we link to the standards we're referring to here or atleast our MIP on `standards-based`
```suggestion
- **Extensibility** - The Multichain API can be integrated with
  [interoperability Snaps](https://snaps.metamask.io/explore/), providing a standards-based interface
  to connect the wallet with dApps on non-EVM networks.
```

- **Seamless multichain UX** - The Multichain API offers improvements over EIP-1193 and [wallet-standard](https://github.com/wallet-standard/wallet-standard) interfaces.

>   It allows dapps to create unified multichain wallet connection flows, trigger transactions across disparate networks, and clearly interpret chain-specific addresses.

Reviewer 4: ```suggestion
  It allows dapps to create unified cross-ecosystem multichain wallet connection flows, trigger transactions across different networks more seamlessly, and clearly interpret chain-specific addresses.
```

  different wallet APIs and SDKs, making it easier for dapps to integrate with wallets, discover
  capabilities, and negotiate interfaces.

## Technical overview

The Multichain API follows the [CAIP-25](https://github.com/ChainAgnostic/CAIPs/blob/main/CAIPs/caip-25.md)
standard for dapps to interface with multichain wallets.
The API includes a method [`wallet_createSession`](https://github.com/ChainAgnostic/CAIPs/blob/main/CAIPs/caip-25.md)
that dapps can call to create a multichain session with a wallet, with specified properties and
authorization scopes.
The session can be created with or without a `sessionId`, and can be updated using the same method
`wallet_createSession`.

Dapps can use the method [`wallet_invokeMethod`](https://github.com/ChainAgnostic/CAIPs/blob/main/CAIPs/caip-27.md)
to call a subset of the [Wallet JSON-RPC API methods](/wallet/reference/json-rpc-api) on a specified chain.
Dapps can use [`wallet_getSession`](https://github.com/ChainAgnostic/CAIPs/blob/main/CAIPs/caip-312.md)
to get the scope of the current session, and
[`wallet_revokeSession`](https://github.com/ChainAgnostic/CAIPs/blob/main/CAIPs/caip-285.md) to
revoke the current session.

> The API also supports [events](../reference/multichain-api-events.md), allowing wallets to notify

Reviewer 2: At least for the `wallet_sessionChanged` I think it would help to provide a high-level overview of the method right here rather than pointing them to the multichain-api-events page. Mainly because we are referencing that event in the diagram.
